### DISCUSSION TITLE
Missing Validators

### DISCUSSION THREAD

**Discussion by aflatoon2874** [prismaneui/prismane#65] [Q&A] [answered]
The list of validators seems to be not comprehensive enough. Where are validators for the following:
- Number validators like `min`, `max`, `multipleOf`, `<`, `>`, `<=`, `>=`, etc.
- Date validators
- RegEx validator for pattern matching. The `username` validator lacks functionality, it should be able to match against a pattern.

**Comment by spleafy** [comment]
Hi! Thanks for the feedback! I will make sure to fix the provided validators, make them more flexible, and most surely introduce the provided validators with the next minor version of Prismane. 

Do you have any other validators in mind that you think will be a great addition to the ones that you listed?

**Reply by aflatoon2874** [reply]
Thanks for a prompt response. I have not started using the library yet. I went through the docs, not in detail though. I must say it is really good effort. Though one thing I didn't like is the abbreviated standard CSS component props. It will take time to get familiar with.

As far as the validators is concerned, yes there is one more validator required that should allow a  custom function to be executed that returns either a message or `null` with a signature like `validate: (v) => {}`

**Reply by spleafy** [reply]
Thank you very much! I hope you like Prismane. 
Could you please give me a use case of the validate function that you suggested to be added? I would really appreciate that.

**Reply by aflatoon2874** [reply]
Sure. Here we go
1. For numbers
```
validate: {
  integer: (v) => Number.isInteger(v) ? null : 'Value must be an integer',
  multipleOf: (v) => v % 10 === 0 ? null : 'Value must be multiple of 10'
}
```
2. For validating dates
```
                    validate: {
                      dateLessThanStartDate: (value) => {
                        if (startDate && value) {
                          const s = new Date(startDate).getTime()
                          const e = new Date(value).getTime()
                          return e < s ? 'End date must be after the Start date' : null
                        }
                        return null
                      },
                    }
```
Hope this helps

**Reply by spleafy** [reply]
I get your point now. Thank you very much! This function is included in the useForm hook. When the hook validates the form, it runs each validator and if one returns a message, the function returns the error. The hook's documentation page should be updated to make that clear.

I am not sure what will be the use case of extracting that functionality from the hook to a separate function.

Thank you for giving those recommendations I will make sure that the validators listed and many more and included in the next minor version!

**Reply by aflatoon2874** [reply]
"I am not sure what will be the use case of extracting that functionality from the hook to a separate function."
I think it makes sense to expose it as a field validator. The user will come to know the error when he/she either clicks submit button or presses enter key. Let's take another use case where I have to instantaneously validate the entered value against an API call, say to check username entered is already taken and to suggest similar available names. In such a scenario it does not make sense to validate by submitting the form again and again.

**Reply by spleafy** [reply]
I understand. Yes, it makes sense to expose it as a form validator. I will add that functionality :)

**Reply by spleafy** [reply]
Thank you for your time and contribution, it means a lot! If anything else comes to mind, please feel free to tell me.

**Comment by spleafy** [comment, marked as answer]
Hi! The new minor version of Prismane ([v1.2.0](https://github.com/prismaneui/prismane/releases/tag/v1.2.0)) introduced 21 new validators, please update to use them!
